File: src/algorithms/matrices/search/island-count/sources/IslandCount.java
// Island Count
// Count the number of islands (connected groups of 1s) in a binary matrix using DFS flood fill.
// An island is a group of adjacent 1s connected horizontally or vertically.
// Time: O(m x n) — every cell is visited at most once
// Space: O(m x n) — DFS call stack depth in the worst case

public class IslandCount {

    public static int islandCount(int[][] grid) {
        int rowCount = grid.length; // @step:initialize
        int colCount = rowCount > 0 ? grid[0].length : 0; // @step:initialize
        int islandTotal = 0; // @step:initialize

        for (int rowIdx = 0; rowIdx < rowCount; rowIdx++) {
            for (int colIdx = 0; colIdx < colCount; colIdx++) {
                if (grid[rowIdx][colIdx] == 1) { // @step:compare-cell
                    islandTotal++; // @step:mark-found
                    dfsFloodFill(grid, rowIdx, colIdx, rowCount, colCount); // @step:mark-found
                }
            }
        }

        return islandTotal; // @step:complete
    }

    private static void dfsFloodFill(int[][] grid, int rowIdx, int colIdx, int rowCount, int colCount) {
        if (rowIdx < 0 || rowIdx >= rowCount) return; // @step:compare-cell
        if (colIdx < 0 || colIdx >= colCount) return; // @step:compare-cell
        if (grid[rowIdx][colIdx] != 1) return; // @step:compare-cell

        grid[rowIdx][colIdx] = 0; // @step:mark-found
        dfsFloodFill(grid, rowIdx - 1, colIdx, rowCount, colCount); // @step:mark-found
        dfsFloodFill(grid, rowIdx + 1, colIdx, rowCount, colCount); // @step:mark-found
        dfsFloodFill(grid, rowIdx, colIdx - 1, rowCount, colCount); // @step:mark-found
        dfsFloodFill(grid, rowIdx, colIdx + 1, rowCount, colCount); // @step:mark-found
    }
}

File: src/algorithms/matrices/search/island-count/sources/island-count.rs
// Island Count
// Count the number of islands (connected groups of 1s) in a binary matrix using DFS flood fill.
// An island is a group of adjacent 1s connected horizontally or vertically.
// Time: O(m × n) — every cell is visited at most once
// Space: O(m × n) — DFS call stack depth in the worst case

fn island_count(grid: &mut Vec<Vec<i32>>) -> i32 {
    let row_count = grid.len(); // @step:initialize
    let col_count = if row_count > 0 { grid[0].len() } else { 0 }; // @step:initialize
    let mut island_total: i32 = 0; // @step:initialize

    for row_idx in 0..row_count {
        for col_idx in 0..col_count {
            if grid[row_idx][col_idx] == 1 {
                // @step:compare-cell
                island_total += 1; // @step:mark-found
                dfs_flood_fill(grid, row_idx, col_idx, row_count, col_count); // @step:mark-found
            }
        }
    }

    island_total // @step:complete
}

fn dfs_flood_fill(
    grid: &mut Vec<Vec<i32>>,
    row_idx: usize,
    col_idx: usize,
    row_count: usize,
    col_count: usize,
) {
    if row_idx >= row_count { return; } // @step:compare-cell
    if col_idx >= col_count { return; } // @step:compare-cell
    if grid[row_idx][col_idx] != 1 { return; } // @step:compare-cell

    grid[row_idx][col_idx] = 0; // @step:mark-found
    if row_idx > 0 { dfs_flood_fill(grid, row_idx - 1, col_idx, row_count, col_count); } // @step:mark-found
    dfs_flood_fill(grid, row_idx + 1, col_idx, row_count, col_count); // @step:mark-found
    if col_idx > 0 { dfs_flood_fill(grid, row_idx, col_idx - 1, row_count, col_count); } // @step:mark-found
    dfs_flood_fill(grid, row_idx, col_idx + 1, row_count, col_count); // @step:mark-found
}
